hwtracer/src/decode/ykpt/mod.rs

```diff
@@ -19,29 +38,232 @@ impl TraceDecoder for YkPTTraceDecoder {
+                }
+            }
+
+            // If we get here, there were no further calls to follow in the block, so we consult
+            // the static successor information.
+            match ent.value.successor() {
+                SuccessorKind::Unconditional { target } => {
+                    if let Some(target_off) = target {
+                        self.cur_loc = ObjLoc::MainObj(*target_off);
+                        return Ok(Some(self.lookup_block_from_main_bin_offset(*target_off)));
+                    } else {
+                        // Divergent control flow.
+                        todo!();
+                    }
+                }
+                SuccessorKind::Conditional {
+                    taken_target,
+                    not_taken_target,
+                } => {
+                    // If we don't have any TNT choices buffered, get more.
+                    if self.tnts.is_empty() {
+                        if self.seek_tnt()? == false {
+                            // The packet stream was exhausted.
+                            return Ok(None);
+                        }
+                    }
+                    // Find where to go next based on whether the trace took this branch or not.
+                    let target_off = if self.tnts.pop_front().unwrap() {
+                        *taken_target
+                    } else {
+                        if let Some(ntt) = not_taken_target {
+                            *ntt
+                        } else {
+                            // Divergent control flow.
+                            panic!();
+                        }
+                    };
+                    self.cur_loc = ObjLoc::MainObj(target_off);
+                    return Ok(Some(self.lookup_block_from_main_bin_offset(target_off)));
+                }
+                SuccessorKind::Dynamic => {
+                    // We can only know the successor via a TIP update in a packet. Fall through to
+                    // `self.seek_tip()`.
+                    self.cur_loc = ObjLoc::OtherObjOrUnknown;
+                }
+            }
+        }
+
+        // If we get here then we can't statically figure out where to go next. A packet is going
+        // to have to tell us what to do.
+        self.seek_tip()
+    }
+
+    /// Keep decoding packets until we have some TNT decisions buffered.
+    fn seek_tnt(&mut self) -> Result<bool, HWTracerError> {
+        while self.tnts.is_empty() {
+            let pkt = self.packet()?; // Potentially populates `self.tnts`.
+            if pkt.is_none() {
+                return Ok(false); // Packet stream exhausted.
+            }
+        }
+        Ok(true)
+    }
+
+    /// Keep decoding packets until we receive a TIP update.
+    fn seek_tip(&mut self) -> Result<Option<Block>, HWTracerError> {
+        debug_assert_eq!(self.cur_loc, ObjLoc::OtherObjOrUnknown);
+        loop {
+            if let Some(_) = self.packet()? {
```

Comment:
This should either be `if self.packet()?.is_some()` or a `match` IMHO.

Reply:
Ah, yes, it's a hangover from hacking.

Reply:
Fixed in 390e15a85418d17488860b63e19252b4b05b6e69